src/lib/logical_query_plan/jit_aware_lqp_translator.cpp

```diff
@@ -0,0 +1,273 @@
+  std::unordered_set<std::shared_ptr<AbstractLQPNode>> input_nodes;
+
+  // Traverse query tree until a non-jitable nodes is found in each branch
+  _breadth_first_search(node, [&](auto& current_node) {
+    if (_node_is_jitable(current_node)) {
+      ++num_jittable_nodes;
+      return true;
+    } else {
+      input_nodes.insert(current_node);
+      return false;
+    }
+  });
+
+  // It does not make sense to create a JitOperatorWrapper for fewer than 2 LQP nodes,
+  // but we may want a better heuristic here
+  if (num_jittable_nodes < 2 || input_nodes.size() != 1) {
+    return LQPTranslator::translate_node(node);
+  }
+
+  const auto input_node = *input_nodes.begin();
+
+  auto jit_operator = std::make_shared<JitOperatorWrapper>(translate_node(input_node));
+  auto read_tuple = std::make_shared<JitReadTuple>();
+  jit_operator->add_jit_operator(read_tuple);
+
+  auto filter_node = node;
+  while (filter_node != input_node && filter_node->type() != LQPNodeType::Predicate &&
+         filter_node->type() != LQPNodeType::Union) {
+    filter_node = filter_node->left_input();
+  }
+
+  try {
+    // If we can reach the input node without encountering a UnionNode or PredicateNode,
+    // there is no need to filter any tuples
+    if (filter_node != input_node) {
+      // However, if we do need to filter, we first convert the filter node to a JitExpression ...
+      const auto expression = _translate_to_jit_expression(filter_node, *read_tuple, input_node);
+      // make sure that expression get computed ...
+      jit_operator->add_jit_operator(std::make_shared<JitCompute>(expression));
+      // and then filter on the resulting boolean.
+      jit_operator->add_jit_operator(std::make_shared<JitFilter>(expression->result()));
+    }
+
+    // Identify the top-most projection node to determine the output columns
+    auto top_projection = node;
+    while (top_projection != input_node && top_projection->type() != LQPNodeType::Projection) {
+      top_projection = top_projection->left_input();
+    }
+
+    // Add a compute operator for each output column that computes the column value.
+    auto write_table = std::make_shared<JitWriteTuple>();
+    for (const auto& output_column :
+         boost::combine(top_projection->output_column_names(), top_projection->output_column_references())) {
+      const auto expression = _translate_to_jit_expression(output_column.get<1>(), *read_tuple, input_node);
+      // It the JitExpression is of type ExpressionType::Column, there is no need to add a compute node, since it
+      // would not compute anything anyway
+      if (expression->expression_type() != ExpressionType::Column) {
+        jit_operator->add_jit_operator(std::make_shared<JitCompute>(expression));
+      }
+      write_table->add_output_column(output_column.get<0>(), expression->result());
+    }
+    jit_operator->add_jit_operator(write_table);
+    return jit_operator;
+  } catch (const std::exception& exception) {
```

Comment:
When does this happen? I am not a fan of using exceptions for expected control flow. Can we do this in a way where we check the return values of the methods or would that make things more complicated?

Reply:
I'd second Markus on his

Reply:
This can happen in a number of cases. E.g. if some expression type or operand is not supported, if an unexpected node type is encountered or if the source of a column cannot be located.
We could add additional control flow for all of this, but the checks are quite spread across the code, so I am not sure that would simplify things.

Reply:
Without having looked into it deeply, I think the preferred way should be to make explicit checks for what nodes/expressions are supported and invoke the normal LQPTranslator in a controlled way rather than doing this via an exception. On second thought I am pretty convinced the exception based way is very, very unclean.
std::exception could be anything, e.g., bad_alloc, some important internal Assertion. Just falling back to LQPTranslator would mask away important error messages then.

Reply:
I agree that exceptions are unclean. IMO there are two ways to do explicit checks:
1) do all checks first, decide which translator to use and then do the translation in a subsequent step. This leads to duplicated code, since node types and parameters have to be checked / dispatched twice.
2) do checks and translation in a single step. However, this requires a way to cancel query translation in the middle and exceptions seem to be the cleanest way to achieve that.
I don't really like either option, but cannot think of a good third option. Do you have any ideas?
We could of course switch to a dedicated opossum::translation_exception if that is your primary concern.

Reply:
I think 1. should be the way to go. If some unavoidable code duplication comes out of it that is acceptable, imo. 
If we allow `opossum::translation_exception` in this case we'll have countless `opossum::do_my_control_flow_for_me_exception` within a years time as people adopt this and I think we don't want to get into that position.

Reply:
How about checking the return value for `nullptr`? If something can't be translated, I get a `nullptr` and that way I know to use the non-JIT path. That should keep us from duplicating any code.

Reply:
I don't even think that that would be hacky. It's more of a "please translate this" - "no".

Reply:
Just an example why I don't like exceptions for control flow: You can't debug your program with breakpoint-on-exception because then it breaks in countless places where no error ever occured and you never actually get to step to the actual exception you're debugging because you give up before that. Sure, you could do breakpoint-on-specific-exception-type, but its still a hazard.